src/Functions/parseReadableSize.cpp

```diff
@@ -0,0 +1,322 @@
+#include <base/types.h>
+#include <boost/algorithm/string/case_conv.hpp>
+
+#include <Columns/ColumnNullable.h>
+#include <Columns/ColumnsNumber.h>
+#include <Columns/ColumnString.h>
+#include <Common/Exception.h>
+#include <Common/FunctionDocumentation.h>
+#include <DataTypes/DataTypeNullable.h>
+#include <DataTypes/DataTypesNumber.h>
+#include <Functions/FunctionFactory.h>
+#include <Functions/FunctionHelpers.h>
+#include <Functions/IFunction.h>
+#include <IO/ReadBufferFromString.h>
+#include <IO/ReadHelpers.h>
+#include <cmath>
+#include <string_view>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int BAD_ARGUMENTS;
+    extern const int CANNOT_PARSE_INPUT_ASSERTION_FAILED;
+    extern const int CANNOT_PARSE_NUMBER;
+    extern const int CANNOT_PARSE_TEXT;
+    extern const int ILLEGAL_COLUMN;
+    extern const int UNEXPECTED_DATA_AFTER_PARSED_VALUE;
+}
+
+enum class ErrorHandling : uint8_t
+{
+    Exception,
+    Zero,
+    Null
+};
+
+using ScaleFactors = std::unordered_map<std::string_view, size_t>;
+
+/** parseReadble*Size - Returns the number of bytes corresponding to a given readable binary or decimal size.
+  * Examples:
+  *  - `parseReadableSize('123 MiB')`
+  *  - `parseReadableSize('123 MB')`
+  * Meant to be the inverse of `formatReadable*Size` with the following exceptions:
+  *  - Number of bytes is returned as an unsigned integer amount instead of a float. Decimal points are rounded up to the nearest integer.
+  *  - Negative numbers are not allowed as negative sizes don't make sense.
+  * Flavours:
+  *  - parseReadableSize
+  *  - parseReadableSizeOrNull
+  *  - parseReadableSizeOrZero
+  */
+template <typename Name, ErrorHandling error_handling>
+class FunctionParseReadable : public IFunction
+{
+public:
+    static constexpr auto name = Name::name;
+    static FunctionPtr create(ContextPtr) { return std::make_shared<FunctionParseReadable<Name, error_handling>>(); }
+
+    String getName() const override { return name; }
+    bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return true; }
+    bool useDefaultImplementationForConstants() const override { return true; }
+    size_t getNumberOfArguments() const override { return 1; }
+
+    DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
+    {
+        FunctionArgumentDescriptors args
+        {
+            {"readable_size", static_cast<FunctionArgumentDescriptor::TypeValidator>(&isString), nullptr, "String"},
+        };
+        validateFunctionArgumentTypes(*this, arguments, args);
+        DataTypePtr return_type = std::make_shared<DataTypeUInt64>();
+        return (error_handling == ErrorHandling::Null) ? std::make_shared<DataTypeNullable>(return_type) : return_type;
```

Comment:
Could use `if constexpr` here (though it doesnt' make a performance difference). Also, for slightly more complex code it is nicer to not use a ternary if. So let's do:

```cpp
if constexpr (error_handling == ErrorHandling::Null)
    [...]
else
    [...]
```

Reply:
applied